v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/bigtablechangestreamstopubsub/FailsafeModJsonToPubsubMessageTransformer.java

```diff
@@ -0,0 +1,210 @@
+    public TupleTag<PubsubMessage> transformOut = new TupleTag<>() {};
+
+    /** The tag for the dead letter output of the transformation. */
+    public TupleTag<FailsafeElement<String, String>> transformDeadLetterOut = new TupleTag<>() {};
+
+    private final FailsafeModJsonToPubsubMessageOptions failsafeModJsonToPubsubMessageOptions;
+
+    public FailsafeModJsonToPubsubMessage(
+        PubSubUtils pubSubUtils,
+        FailsafeModJsonToPubsubMessageOptions failsafeModJsonToPubsubMessageOptions) {
+      this.pubSubUtils = pubSubUtils;
+      this.failsafeModJsonToPubsubMessageOptions = failsafeModJsonToPubsubMessageOptions;
+    }
+
+    public PCollectionTuple expand(PCollection<FailsafeElement<String, String>> input) {
+      PCollectionTuple out =
+          input.apply(
+              ParDo.of(
+                      new FailsafeModJsonToPubsubMessageFn(
+                          pubSubUtils, transformOut, transformDeadLetterOut))
+                  .withOutputTags(transformOut, TupleTagList.of(transformDeadLetterOut)));
+      out.get(transformDeadLetterOut).setCoder(failsafeModJsonToPubsubMessageOptions.getCoder());
+      return out;
+    }
+
+    /**
+     * The {@link FailsafeModJsonToPubsubMessageFn} converts a JSON string wrapped in {@link
+     * FailsafeElement} to a {@link PubsubMessage}.
+     */
+    public static class FailsafeModJsonToPubsubMessageFn
+        extends DoFn<FailsafeElement<String, String>, PubsubMessage> {
+      private final PubSubUtils pubSubUtils;
+      public TupleTag<PubsubMessage> transformOut;
+      public TupleTag<FailsafeElement<String, String>> transformDeadLetterOut;
+      private transient Publisher publisher;
+
+      public FailsafeModJsonToPubsubMessageFn(
+          PubSubUtils pubSubUtils,
+          TupleTag<PubsubMessage> transformOut,
+          TupleTag<FailsafeElement<String, String>> transformDeadLetterOut) {
+        this.pubSubUtils = pubSubUtils;
+        this.transformOut = transformOut;
+        this.transformDeadLetterOut = transformDeadLetterOut;
+      }
+
+      @Setup
+      public void setUp() {
+        try {
+          final TopicName projectTopicName =
+              TopicName.of(
+                  pubSubUtils.getDestination().getPubSubProject(),
+                  pubSubUtils.getDestination().getPubSubTopicName());
+          publisher = Publisher.newBuilder(projectTopicName).build();
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+
+      @Teardown
+      public void tearDown() {
+        try {
+          if (publisher != null) {
+            publisher.shutdown();
+            publisher.awaitTermination(5, TimeUnit.MINUTES);
+          }
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) {
+        LOG.info("Reading a failsafeModJsonString");
```

Comment:
This logging will potentially a looot of log lines. Let's make sure we don't log this stuff on per data item basis